src/Analyzers/CSharp/Analyzers/UseNameofInNullableAttribute/CSharpUseNameofInNullableAttributeDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseNameofInNullableAttribute;
+
+/// <summary>
+/// Analyzer that looks for things like `NotNullIfNotNull("param")` and offers to use `NotNullIfNotNull(nameof(param))` instead.
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal sealed class CSharpUseNameofInNullableAttributeDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+{
+    public const string NameKey = nameof(NameKey);
+
+    public CSharpUseNameofInNullableAttributeDiagnosticAnalyzer()
+        : base(IDEDiagnosticIds.UseNameofInNullableAttributeDiagnosticId,
+               EnforceOnBuildValues.UseNameofInNullableAttribute,
+               option: null,
+               new LocalizableResourceString(
+                   nameof(CSharpAnalyzersResources.Use_nameof), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+    {
+    }
+
+    public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+    protected override void InitializeWorker(AnalysisContext context)
+    {
+        context.RegisterCompilationStartAction(context =>
+        {
+            if (context.Compilation.LanguageVersion() >= LanguageVersion.CSharp11)
+                context.RegisterSyntaxNodeAction(AnalyzeAttribute, SyntaxKind.Attribute);
+        });
+    }
+
+    private void AnalyzeAttribute(SyntaxNodeAnalysisContext context)
+    {
+        var cancellationToken = context.CancellationToken;
+        var attribute = (AttributeSyntax)context.Node;
+        var semanticModel = context.SemanticModel;
+
+        if (attribute.ArgumentList is null)
+            return;
+
+        var attributeName = attribute.Name.GetRightmostName()?.Identifier.ValueText;
+        if (attributeName is null)
+            return;
+
+        if (attributeName
+                is not "NotNullIfNotNullAttribute"
+                and not "MemberNotNullAttribute"
+                and not "MemberNotNullWhenAttribute"
+                and not "NotNullIfNotNull"
+                and not "MemberNotNull"
+                and not "MemberNotNullWhen")
```

Comment:
Can we extend this to CallerArgumentExpression?

Reply:
seems reasonable.